Sources/MongoMobile/MongoMobile.swift

```diff
@@ -14,37 +14,43 @@ public struct MongoClientSettings {
     }
 }
 
+/// A MongoDB server error code.
+/// - SeeAlso: https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.err
+public typealias ServerErrorCode = Int
+
 public enum MongoMobileError: LocalizedError {
-    case invalidClient()
-    case invalidInstance(message: String)
-    case invalidLibrary()
-    case instanceDropError(message: String)
-    case cleanupError(message: String)
+    /// Thrown when MongoMobile is incorrectly used.
+    case logicError(message: String)
+
+    /// Thrown when an embedded mongo server instance returns an error.
+    /// - SeeAlso: https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.err
+    case embeddedMongoError(code: ServerErrorCode, message: String)
+
+    /// Thrown when MongoMobile encounters internal errors it cannot recover from.
+    case internalError(message: String)
 
     /// a string to be printed out when the error is thrown
     public var errorDescription: String? {
         switch self {
-        case let .invalidInstance(message),
-            let .instanceDropError(message),
-            let .cleanupError(message):
+        case let .embeddedMongoError(_, message),
+             let .internalError(message),
+             let .logicError(message):
             return message
-        default:
-            return nil
         }
     }
 }
 
-public struct MongoEmbeddedV1Error: LocalizedError {
-    private let statusMessage: String
-    private let error: mongo_embedded_v1_error
-
-    public var errorDescription: String? {
-        return "\(error): \(statusMessage)"
-    }
-
-    internal init(_ error: mongo_embedded_v1_error, statusMessage: String) {
-        self.statusMessage = statusMessage
-        self.error = error
+/// Given an `OpaquePointer` to a `mongo_embedded_v1_status`, get an appropriate `MongoMobileError`.
+internal func parseEmbeddedMongoError(_ status: OpaquePointer?) -> MongoMobileError {
+    let error = mongo_embedded_v1_error(rawValue: mongo_embedded_v1_status_get_error(status))
+
+    switch error {
+    case MONGO_EMBEDDED_V1_ERROR_EXCEPTION:
+        return .embeddedMongoError(code: getStatusCode(status), message: getStatusExplanation(status))
+    case MONGO_EMBEDDED_V1_ERROR_LIBRARY_ALREADY_INITIALIZED, MONGO_EMBEDDED_V1_ERROR_LIBRARY_NOT_INITIALIZED:
+        return .logicError(message: getStatusExplanation(status))
```

Comment:
The current short-circuiting should preclude this code path ever being taken, but I think it's worth it to have in case the code structure ever changes / new paths are added that don't have short circuiting.